src/abstractdataframe/selection.jl

```diff
@@ -978,8 +989,17 @@ ERROR: ArgumentError: column :x in returned data frame is not equal to grouping
 
 See [`select`](@ref) for more examples.
 """
-transform(df::AbstractDataFrame, @nospecialize(args...); copycols::Bool=true, renamecols::Bool=true) =
-    select(df, :, args..., copycols=copycols, renamecols=renamecols)
+function transform(df::AbstractDataFrame, @nospecialize(args...); copycols::Bool=true, renamecols::Bool=true)
+    idx = index(df)
+    newargs = Any[if sel isa Pair{<:ColumnIndex, Symbol}
+            idx[first(sel)] => copy => last(sel)
+        elseif sel isa Pair{<:ColumnIndex, <:AbstractString}
+            idx[first(sel)] => copy => Symbol(last(sel))
+        else
+            sel
+        end for sel in args]
```

Comment:
This is needed for `copycols=false`. For `copycols=true` it is indeed not required as we have `_add_col_check_copy` logic for this, where we check:
```
if copycols && !(fun isa ByRow) && (v isa SubArray || any(i -> vpar === parent(cdf[i]), parent_cols))
```
but this check will fail if we do the rewrite to `copy` so we will not copy twice.

The problem we want to solve for `copycols=false` is (this is before this PR):
```
julia> df = DataFrame(a=1)
df21×1 DataFrame
 Row │ a     
     │ Int64 
─────┼───────
   1 │     1

julia> df2 = transform(df, :a=>:b, copycols=false)
1×2 DataFrame
 Row │ a      b     
     │ Int64  Int64 
─────┼──────────────
   1 │     1      1

julia> df2.a === df2.b === df.a
true
```
Now you can assure yourself that we do not make copy twice like this. On main (after compilation):
```

julia> df = DataFrame(a=1:10^7);

julia> @allocated transform(df, :a => :b)
160005536
```
And the same in this PR:
```
julia> df = DataFrame(a=1:10^7);

julia> @allocated transform(df, :a => :b)
160007344
```
(the difference is because of the additional logic)

I could add an additional `if` and use the new logic only if `copycols=false`, and the old logic if `copycols=true`. What would you prefer?

Reply:
Got it. Maybe just a comment would be useful.

Reply:
I have now pushed a comment. Also I have added a comment in NEWS.md that we will consider not-making a copy in similar situations a bug (if we identify them).